File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.py
import sys

input = sys.stdin.readline
N = int(input())
A = list(map(int, input().split()))


def solution():
    graph = [[] for _ in range(N)]
    dp = [0] * N
    for i in range(1, N):
        now = A[i]
        for j in range(i + 1):
            prev = A[j]
            if now > prev:
                graph[i].append(j)

    for i in range(N):
        l = []
        for prev in graph[i]:
            l.append(dp[prev])
        dp[i] = (max(l) if l else 0) + 1

    print(max(dp))


solution()

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.js
let fs = require('fs');
let [N, s] = fs.readFileSync(0, 'utf8').trim().split('\n');
N = +N;
s = s.split(' ').map((v) => +v);

function solution(N, s) {
  const graph = new Array(N).fill(0).map(() => []);

  for (let i = 1; i < N; i++) {
    for (let j = 0; j < i; j++) {
      if (s[j] < s[i]) graph[i].push(j);
    }
  }

  const dp = new Array(N).fill(1);
  for (let i = 1; i < N; i++) {
    let max = 0;
    graph[i].forEach((idx) => {
      if (dp[idx] > max) max = dp[idx];
    });
    dp[i] = max + 1;
  }

  return Math.max(...dp);
}

console.log(solution(N, s));
